### DISCUSSION TITLE
C++ help

### DISCUSSION THREAD

**Discussion by ashishdhole** [idaholab/moose#27431] [Q&A Modules: Phase field]
Hello, 
I want to add a condition in my kernel where residual of ACGrGrPoly > residual of ACInterface (i.e. driving pressure > gradient energy density). I am very new to C++ and not able to get how to do it.
I have tried
```
Real interRes = (_grad_u[_qp] * _kappa[_qp] * _L[_qp] * _grad_test[_i][_qp]);
if (std::abs((_L[_qp] * _mu[_qp] * (op * op * op - op + 3.0 * op * SumOPj)) > std::abs(interRes))
```
but I am not sure if this is how it is done. Can you please help me. The residual of ACInterface has _L[_qp] term which is not in the ACGrGrPoly. can you please help. Thank you

**Comment by GiudGiud** [comment]
hello

comparing two numbers like this in C++ is fine.
However, 
```
(_grad_u[_qp] * _kappa[_qp] * _L[_qp] * _grad_test[_i][_qp]);
```
looks like the residual of a diffusion-like operator so it seems correct.
however
```
std::abs((_L[_qp] * _mu[_qp] * (op * op * op - op + 3.0 * op * SumOPj)
```
does not look like a weak form residual? Unless the 'test' function is implicitly in one of the terms?

Guillaume

**Reply by ashishdhole** [reply]
the second part is the residual of ACGrGrPoly, which has test function in offdiagonaljacobian and not in residual part. do I have to add the test function in my relation?

**Reply by GiudGiud** [reply]
ok I am not familiar with this kernel but I dot think the residual does not have the test function

I looked up the base classes and I found
```
template <typename T>
Real
ACBulk<T>::precomputeQpResidual()
{
  // Get free energy derivative from function
  Real dFdop = computeDFDOP(Residual);

  // Set residual
  return _L[_qp] * dFdop;
}
```

the output of `precomputeQpResidual()` from a KernelValue class is multiplied by the test function.
so you should instead multiply by the test function

**Reply by ashishdhole** [reply]
If I put test function in the relation, the make happens successfully, but when running input file, it terminates the calculations at time=0. but if I am not putting this condition, my input file runs perfectly fine. Not sure what exactly is happening.

**Reply by GiudGiud** [reply]
how does it terminate? through non-convergence or through a crash / seg fault?

**Reply by ashishdhole** [reply]
segmentation fault

**Reply by GiudGiud** [reply]
you ll want to recompile in debug mode and re-run. This may give you a backtrace that tells you why it seg faults
if not, follow these instructions to obtain that information
https://mooseframework.inl.gov/application_development/debugging.html
